### PR TITLE
Add zone aware replication support

### LINKED ISSUE #4600: Asynchronous Multi Site Ingester Replication
We are trying to deploy on-premise Loki architecture in 2 data centers and we want our deployment to be resilient to site-level malfunctions.

We configured our Promtail agents across all data centers to send logs to a single centralized Loki deployment hosted in a single datacenter. 

We want to deploy a backup Loki on a different site that we will use instead of the main one at the time of an outage. We use Cassandra for both chunk and index store and we can easily replicate Cassandra across 2 data centers so that data it is available in the backup site when needed, however data in the "main" Loki stored in the ingesters won't be available to the backup site until it is flushed to the backend storage.

Now we understand that the data won't truly be lost due to the WAL, but it can lead to the lost of important log lines, originating from the site and going into the site, that might be extremely useful when figuring out the root cause of the outage and might hold back application hosted on the failed datacenter from getting a clear picture of the state of their application. Potentially a few hours of logs might be unavailable to us at the time of an outage despite the effort we put into setting up a backup deployment. 

We haven't fully thought about the technical details of the solution, but we think that providing a set of ingesters a way to "fetch" for themselves chunks stored on a remote set of ingesters can increase Loki's resilience. We think that a mechanism to prevent ingesters from flushing the same set of data to the backend storage when the sites are up and especially when the master site is down is key to this solution and would be challenging to execute.

An alternative solution would be to provide a set of ingesters with some kind of a site identifier. In that case queriers would turn to multiple sites to fetch logs, depending on the query, and a single site outage would lead to the lose of logs originating only from that site. We realize this solution introduces some complications, a set of ingesters and distributors would have to be deployed in each site and fetching logs might be performed in inconsistent latencies.

Any help or feedback would be greatly appreciated.

### PR DESCRIPTION
**Pull request by taisho6339**
**What this PR does / why we need it**:

To support zone aware replication, I've added some codes and documents.
Cortex already supports this feature and Loki uses same codes for resolving ingester instances so I think we can already use this feature in also Loki.

In addition, lifecycler and ring config is exactly same as Cortex's

**Which issue(s) this PR fixes**:
Fixes #4600

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [ ] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

**Comment by taisho6339** [comment]
check-generated-files step looks broken in CI.

File: docs/sources/configuration/_index.md [hunk 1]
```diff
@@ -1006,6 +1006,11 @@ lifecycler:
     # CLI flag: -distributor.replication-factor
     [replication_factor: <int> | default = 3]
 
+    # True to enable the zone-awareness and replicate ingested samples across
```
Comment (GrafanaWriter) [inline docs/sources/configuration/_index.md:1009]: minor nit to further clarify
```suggestion
    # Set to true to enable zone-awareness and replicate ingested samples across
```

**Review by GrafanaWriter — APPROVED** [review]
Minor edit to clarify.

**Comment by taisho6339** [comment]
apparently, these changes have already merged by another PR so I'll close it.

**Pull request closed without merging**